Plugin/Framework/Search/Request/AbstractHandler.php

```diff
@@ -0,0 +1,304 @@
+use Nosto\Cmp\Helper\SearchEngine;
+use Nosto\Cmp\Logger\LoggerInterface;
+use Nosto\Cmp\Model\Filter\FiltersInterface;
+use Nosto\Cmp\Model\Service\Recommendation\StateAwareCategoryServiceInterface;
+use Nosto\Cmp\Plugin\Catalog\Block\ParameterResolverInterface;
+use Nosto\Cmp\Utils\CategoryMerchandising;
+use Nosto\Cmp\Utils\Search;
+use Nosto\Tagging\Helper\Account as NostoHelperAccount;
+
+abstract class AbstractHandler
+{
+    const KEY_ES_PRODUCT_ID = '_id';
+    const KEY_MYSQL_PRODUCT_ID = 'entity_id';
+    const KEY_BIND_TO_QUERY = 'catalog_view_container';
+    const KEY_BIND_TO_GRAPHQL = 'graphql_product_search';
+    const KEY_CATEGORY_FILTER = 'category_filter';
+    const KEY_QUERIES = 'queries';
+    const KEY_FILTERS = 'filters';
+    const KEY_VALUE = 'value';
+    const KEY_CMP = 'nosto_cmp_id_search';
+    const KEY_RESULTS_FROM = 'from';
+    const KEY_RESULT_SIZE = 'size';
+
+    /**
+     * @var ParameterResolverInterface
+     */
+    private $parameterResolver;
+
+    /**
+     * @var LoggerInterface
+     */
+    private $logger;
+
+    /**
+     * @var SearchEngine
+     */
+    private $searchEngineHelper;
+
+    /**
+     * @var StoreManagerInterface
+     */
+    protected $storeManager;
+
+    /**
+     * @var NostoHelperAccount
+     */
+    private $accountHelper;
+
+    /**
+     * @var StateAwareCategoryServiceInterface
+     */
+    protected $categoryService;
+
+    /**
+     * @param ParameterResolverInterface $parameterResolver
+     * @param SearchEngine $searchEngineHelper
+     * @param StoreManagerInterface $storeManager
+     * @param NostoHelperAccount $nostoHelperAccount
+     * @param StateAwareCategoryServiceInterface $categoryService
+     * @param LoggerInterface $logger
+     */
+    public function __construct(
+        ParameterResolverInterface $parameterResolver,
+        SearchEngine $searchEngineHelper,
+        StoreManagerInterface $storeManager,
+        NostoHelperAccount $nostoHelperAccount,
+        StateAwareCategoryServiceInterface $categoryService,
+        LoggerInterface $logger
+    ) {
+        $this->parameterResolver = $parameterResolver;
+        $this->logger = $logger;
+        $this->searchEngineHelper = $searchEngineHelper;
+        $this->storeManager = $storeManager;
+        $this->accountHelper = $nostoHelperAccount;
+        $this->categoryService = $categoryService;
+    }
+
+    /**
+     * @param array $requestData
+     * @return void
+     */
+    public function handle(array &$requestData)
+    {
+        $this->logger->debugCmp(
+            sprintf(
+                'Using %s as search engine',
+                $this->searchEngineHelper->getCurrentEngine()
+            ),
+            $this
+        );
+        $this->preFetchOps($requestData);
+        $productIds = $this->getCmpProductIds(
+            $this->parsePageNumber($requestData),
+            $this->parseLimit($requestData)
+        );
+        $this->cleanUpCmpSort($requestData);
```

Comment:
This must be before `$this->getCmpProductIds`. If getting CMP results fails / throws an exception the cmp sort is not removed from the ES request data